### PR TITLE
MGDAPI-5152 - replace deprecated clock package usage

### PR DESCRIPTION
**Pull request by adam-cattermole**
[MGDAPI-5152](https://issues.redhat.com/browse/MGDAPI-5152)

As part of the changes for the above ticket for openshift v4.13 support I was unable to resolve dep conflicts in 3scale-operator without removing use of the deprecated clock package. I've swapped the two uses for the newer version in this PR.

### CODE REVIEW

**Comment by austincunningham** [comment]
@adam-cattermole  code changes look fine, will see if the tests pass before merging

File: go.mod [hunk 1]
```diff
@@ -23,9 +23,10 @@ require (
 	github.com/stretchr/testify v1.8.0
 	golang.org/x/mod v0.6.0-dev.0.20220419223038-86c51ed26bb4
 	gopkg.in/yaml.v2 v2.4.0
-	k8s.io/api v0.24.3
-	k8s.io/apimachinery v0.24.3
-	k8s.io/client-go v0.24.3
+	k8s.io/api v0.25.0
```
[the remark this answers is not in the document]
Comment (eguzki) [reply go.mod:26]: 0.25.0 requires go 1.19 https://github.com/kubernetes/api/blob/v0.25.0/go.mod

**Comment by adam-cattermole** [comment]
Added a new commit [d90e71a](https://github.com/3scale/3scale-operator/pull/805/commits/d90e71a75429033c8de6a83dbcec11a71671fd0d) as I discovered that we are unable to successfully deploy on OSD v4.13.

On v4.11 when updating any of the deployment configs a new replication controller is created and there are warnings about a duplicate `FILE_UPLOAD_STORAGE` env var in the deploy logs. However on v4.13 these no longer show as warnings but as errors. This duplicate env var is present as it is added from the base config map and then appended again.

*v4.11.32*
```
--> pre: Running hook pod ...
NOTE: Gem::Specification#has_rdoc= is deprecated with no replacement. It will be removed on or after 2018-12-01.
Gem::Specification#has_rdoc= called from /opt/deps/rubygems/github.com/3scale/prawn/prawn-external-gitcommit-88aead0d97e230d08e9f51a18711d01317965bfe/app/prawn-core.gemspec:36.
fatal: not a git repository (or any of the parent directories): .git
Removing :environment prerequisite from db:create
I, [2023-03-31T10:45:53.839087 #1] INFO -- : ActiveMerchant MODE set to 'production'
W, [2023-03-31T10:45:53.854215 #1] WARN -- [Bugsnag]: No valid API key has been set, notifications will not be sent
I, [2023-03-31T10:45:54.026795 #1] INFO -- : [Core] Using http://backend-listener:3000/internal/ as URL
OpenIdAuthentication.store is nil. Using in-memory store.
Creating scope :admins. Overwriting existing method User.admins.
Creating scope :by_name. Overwriting existing method Cinstance.by_name.
[core] non-native log levels verbose, notice, critical emulated using UNKNOWN severity
Backend Internal API version 3.4.3 status: ok
Connected to mysql2://root@system-mysql/system
Connected to redis://system-redis:6379/1
--> pre: Success
W0331 10:45:59.726113 1 warnings.go:70] would violate PodSecurity "restricted:v1.24": allowPrivilegeEscalation != false (containers "system-master", "system-provider", "system-developer" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (containers "system-master", "system-provider", "system-developer" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or containers "system-master", "system-provider", "system-developer" must set securityContext.runAsNonRoot=true), seccompProfile (pod or containers "system-master", "system-provider", "system-developer" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
W0331 10:45:59.754015 1 warnings.go:70] would violate PodSecurity "restricted:v1.24": allowPrivilegeEscalation != false (containers "system-master", "system-provider", "system-developer" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (containers "system-master", "system-provider", "system-developer" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or containers "system-master", "system-provider", "system-developer" must set securityContext.runAsNonRoot=true), seccompProfile (pod or containers "system-master", "system-provider", "system-developer" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
--> Scaling up system-app-2 from 0 to 1, scaling down system-app-1 from 1 to 0 (keep 1 pods available, don't exceed 2 pods)
Scaling system-app-2 up to 1
W0331 10:45:59.764889 1 warnings.go:70] spec.template.spec.containers[0].env[47].name: duplicate name "FILE_UPLOAD_STORAGE"
W0331 10:45:59.764905 1 warnings.go:70] spec.template.spec.containers[1].env[47].name: duplicate name "FILE_UPLOAD_STORAGE"
W0331 10:45:59.764909 1 warnings.go:70] spec.template.spec.containers[2].env[47].name: duplicate name "FILE_UPLOAD_STORAGE"
Scaling system-app-1 down to 0
W0331 10:47:05.746333 1 warnings.go:70] spec.template.spec.containers[0].env[47].name: duplicate name "FILE_UPLOAD_STORAGE"
W0331 10:47:05.746347 1 warnings.go:70] spec.template.spec.containers[1].env[47].name: duplicate name "FILE_UPLOAD_STORAGE"
W0331 10:47:05.746351 1 warnings.go:70] spec.template.spec.containers[2].env[47].name: duplicate name "FILE_UPLOAD_STORAGE"
W0331 10:47:06.778809 1 warnings.go:70] would violate PodSecurity "restricted:v1.24": allowPrivilegeEscalation != false (containers "system-master", "system-provider", "system-developer" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (containers "system-master", "system-provider", "system-developer" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or containers "system-master", "system-provider", "system-developer" must set securityContext.runAsNonRoot=true), seccompProfile (pod or containers "system-master", "system-provider", "system-developer" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
--> post: Running hook pod ...
NOTE: Gem::Specification#has_rdoc= is deprecated with no replacement. It will be removed on or after 2018-12-01.
Gem::Specification#has_rdoc= called from /opt/deps/rubygems/github.com/3scale/prawn/prawn-external-gitcommit-88aead0d97e230d08e9f51a18711d01317965bfe/app/prawn-core.gemspec:36.
fatal: not a git repository (or any of the parent directories): .git
Removing :environment prerequisite from db:create
I, [2023-03-31T10:47:18.526897 #1] INFO -- : ActiveMerchant MODE set to 'production'
W, [2023-03-31T10:47:18.541539 #1] WARN -- [Bugsnag]: No valid API key has been set, notifications will not be sent
I, [2023-03-31T10:47:18.711720 #1] INFO -- : [Core] Using http://backend-listener:3000/internal/ as URL
OpenIdAuthentication.store is nil. Using in-memory store.
Creating scope :admins. Overwriting existing method User.admins.
Creating scope :by_name. Overwriting existing method Cinstance.by_name.
[core] non-native log levels verbose, notice, critical emulated using UNKNOWN severity
Backend Internal API version 3.4.3 status: ok
Connected to mysql2://root@system-mysql/system
Connected to redis://system-redis:6379/1
Enqueued BackendStorageRewriteWorker#5b9bbe78ad70d110a38d8a61 with args: [1]
Enqueued BackendStorageRewriteWorker#318f39ded0f3d1e70cca4165 with args: [2]
Enqueued 2 accounts for rewrite
--> post: Success
--> Success
```
*v4.13.0-rc.0*
```
--> pre: Running hook pod ...
NOTE: Gem::Specification#has_rdoc= is deprecated with no replacement. It will be removed on or after 2018-12-01.
Gem::Specification#has_rdoc= called from /opt/deps/rubygems/github.com/3scale/prawn/prawn-external-gitcommit-88aead0d97e230d08e9f51a18711d01317965bfe/app/prawn-core.gemspec:36.
fatal: not a git repository (or any of the parent directories): .git
Removing :environment prerequisite from db:create
I, [2023-03-31T12:23:43.046402 #1] INFO -- : ActiveMerchant MODE set to 'production'
W, [2023-03-31T12:23:43.061022 #1] WARN -- [Bugsnag]: No valid API key has been set, notifications will not be sent
I, [2023-03-31T12:23:43.219970 #1] INFO -- : [Core] Using http://backend-listener:3000/internal/ as URL
OpenIdAuthentication.store is nil. Using in-memory store.
Creating scope :admins. Overwriting existing method User.admins.
Creating scope :by_name. Overwriting existing method Cinstance.by_name.
[core] non-native log levels verbose, notice, critical emulated using UNKNOWN severity
Backend Internal API version 3.4.3 status: ok
Connected to mysql2://root@system-mysql/system
Connected to redis://system-redis:6379/1
--> pre: Success
error: couldn't assign source annotation to deployment system-app-2: failed to create manager for existing fields: failed to convert new object (3scale-operator/system-app-2; /v1, Kind=ReplicationController) to smd typed: errors:
.spec.template.spec.containers[name="system-master"].env: duplicate entries for key [name="FILE_UPLOAD_STORAGE"]
.spec.template.spec.containers[name="system-provider"].env: duplicate entries for key [name="FILE_UPLOAD_STORAGE"]
.spec.template.spec.containers[name="system-developer"].env: duplicate entries for key [name="FILE_UPLOAD_STORAGE"]
```

With the changes in the latest commit we no longer have the duplicate entry and the deployment config can be updated successfully on v4.13:

```
--> pre: Running hook pod ...
Removing :environment prerequisite from db:create
I, [2023-03-31T13:36:08.725595 #1] INFO -- : ActiveMerchant MODE set to 'production'
W, [2023-03-31T13:36:08.743611 #1] WARN -- [Bugsnag]: No valid API key has been set, notifications will not be sent
I, [2023-03-31T13:36:08.918910 #1] INFO -- : [Core] Using http://backend-listener:3000/internal/ as URL
OpenIdAuthentication.store is nil. Using in-memory store.
Creating scope :admins. Overwriting existing method User.admins.
Creating scope :by_name. Overwriting existing method Cinstance.by_name.
[core] non-native log levels verbose, notice, critical emulated using UNKNOWN severity
Backend Internal API version 3.4.3 status: ok
Connected to mysql2://root@system-mysql/system
Connected to redis://system-redis:6379/1
--> pre: Success
W0331 13:36:13.598613 1 warnings.go:70] would violate PodSecurity "restricted:latest": seccompProfile (pod or containers "system-master", "system-provider", "system-developer" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
W0331 13:36:13.641820 1 warnings.go:70] would violate PodSecurity "restricted:latest": seccompProfile (pod or containers "system-master", "system-provider", "system-developer" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
--> Scaling up system-app-2 from 0 to 1, scaling down system-app-1 from 1 to 0 (keep 1 pods available, don't exceed 2 pods)
Scaling system-app-2 up to 1
Scaling system-app-1 down to 0
W0331 13:37:15.742399 1 warnings.go:70] would violate PodSecurity "restricted:latest": seccompProfile (pod or containers "system-master", "system-provider", "system-developer" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
--> post: Running hook pod ...
Removing :environment prerequisite from db:create
I, [2023-03-31T13:37:24.713843 #1] INFO -- : ActiveMerchant MODE set to 'production'
W, [2023-03-31T13:37:24.731184 #1] WARN -- [Bugsnag]: No valid API key has been set, notifications will not be sent
I, [2023-03-31T13:37:24.907571 #1] INFO -- : [Core] Using http://backend-listener:3000/internal/ as URL
OpenIdAuthentication.store is nil. Using in-memory store.
Creating scope :admins. Overwriting existing method User.admins.
Creating scope :by_name. Overwriting existing method Cinstance.by_name.
[core] non-native log levels verbose, notice, critical emulated using UNKNOWN severity
Backend Internal API version 3.4.3 status: ok
Connected to mysql2://root@system-mysql/system
Connected to redis://system-redis:6379/1
Enqueued BackendStorageRewriteWorker#ec164faac0d81388c0fdd09f with args: [1]
Enqueued BackendStorageRewriteWorker#23d2aaff28bd257b1638c692 with args: [2]
Enqueued 2 accounts for rewrite
--> post: Success
--> Success
```

**Comment by valerymo** [comment]
It's tested with Rhoam, please see my last comment in https://github.com/integr8ly/integreatly-operator/pull/3148   
- Install of 3scale and Rhoam was successful

File: pkg/3scale/amp/component/system.go [hunk 2]
```diff
@@ -243,7 +243,6 @@ func (system *System) buildSystemBaseEnv() []v1.EnvVar {
 
 	if system.Options.S3FileStorageOptions != nil {
 		result = append(result,
-			helper.EnvVarFromConfigMap("FILE_UPLOAD_STORAGE", "system-environment", "FILE_UPLOAD_STORAGE"),
```
Comment (eguzki) [inline pkg/3scale/amp/component/system.go:246]: Thanks for spotting this issue. I have checked history and it has been there for quite a long time. 

We should not add this change in this PR. It deserves a its own PR.

**Review by eguzki — CHANGES_REQUESTED** [review]

File: pkg/3scale/amp/component/system.go (hunk 2 above)
Comment (adam-cattermole) [reply pkg/3scale/amp/component/system.go]: Opened a new PR for this change (#811) and removed the commit.

**Comment by eguzki** [comment]
I do not know why, circleCI is not running in your PRs from forked repo

```
Could not find a usable config.yml, you may have revoked the CircleCI OAuth app.
```
:shrug:

**Comment by eguzki** [comment]
It's weird, because from my "second" account (with no permissions), my PR runs circleCI tests https://github.com/3scale/3scale-operator/pull/812 :woman_shrugging:

**Comment by adam-cattermole** [comment]
@eguzki Looks like they're running now - I had not authorized the CircleCI GitHub app

**Comment by adam-cattermole** [comment]
Triggered once more as they were running in my own circle-ci workspace instead of the 3scale one. [Looks good now](https://app.circleci.com/pipelines/github/3scale/3scale-operator?branch=pull%2F805) although the check on this PR is pointing to the wrong pipeline (5039 instead of 5044, probably because the commit didn't change)
